ftw/zipextract/zipextracter.py

```diff
@@ -0,0 +1,289 @@
+        return file_list
+
+
+class ZipExtracter():
+    """
+    This should be initialized with a file object.
+    It implements an "extract" method, which creates
+    a new file object for every file in the zip, so
+    it needs to extract that file and save it to disk,
+    but also create the file object in zope
+    """
+
+    def __init__(self, context, max_size=None):
+        ifile = IFile(context)
+        if not ifile.is_zip():
+            err_msg = "{} is not a zip file, can't extract from it.".format(
+                context.absolute_url_path())
+            raise TypeError(err_msg)
+        file = ifile.get_blob()
+        self.context = context
+        self.parent_node = context.getParentNode()
+        self.zfile = zipfile.ZipFile(file.open())
+        self.file_name = os.path.basename(self.context.absolute_url_path())
+        self.file_name = os.path.splitext(self.file_name)[0]
+        self.parent_folder = os.path.dirname(self.context.absolute_url_path())
+        self.file_infos = self.extract_file_infos()
+        self._extract_file_tree()
+        self.max_size = max_size
+        self.file_creator = IFileCreator(self.context)
+
+    def _isdir(self, type):
+        if type == "Folder":
+            return True
+        return False
+
+    def extract_file_infos(self):
+        return self.zfile.infolist()
+
+    @staticmethod
+    def generate_id(name, context):
+        normalizer = component.getUtility(IIDNormalizer)
+        chooser = INameChooser(context)
+        normalized = normalizer.normalize(name)
+        normalized = normalized.replace('_', '-').replace(' ', '-').lower()
+        return chooser.chooseName(normalized, context)
+
+    @staticmethod
+    def generate_dict_key(name, d):
+        name = os.path.splitext(name)[0]
+        if name not in d:
+            return name
+        i = 1
+        while True:
+            new_name = name + "-{}".format(i)
+            if new_name in d:
+                i += 1
+                continue
+            return new_name
+
+    def _extract_file_tree(self):
+        self.file_tree = _FolderNode(None, "")
+        # We first make all the directories
+        for info in self.file_infos:
+            is_folder = self.is_folder(info.filename)
+            target_path = self.get_target_path(info)
+            keys = target_path.split(os.path.sep)
+            if not is_folder:
+                keys = keys[:-1]
+            curr_node = self.file_tree
+            for k in keys:
+                if k not in curr_node.subtree:
+                    curr_node.subtree[k] = _FolderNode(curr_node, k)
+                curr_node = curr_node.subtree[k]
+        # Now we add the files
+        for info in self.file_infos:
+            is_folder = self.is_folder(info.filename)
+            target_path = self.get_target_path(info)
+            if is_folder:
+                continue
+            keys = target_path.split(os.path.sep)
+            curr_node = self.file_tree
```

Comment:
This method is quite long. I'd try to refactor it by extracting functionality into other methods.
I'd also try to only have one `for`-loop over the file infos.

Reply:
The two `for` loops were there because you can have files with the same name as folders inside zip files. But with how the tree is implemented now, this is not needed anymore. This shortens the method a lot, let me know if you still think I should try to delegate some functionality to other methods.